### ISSUE TITLE
please add NodeJS project w/ a dependency that has native module

### ISSUE THREAD

**Issue by Dmitry-N-Medvedev** [nanovms/ops-examples#40]
I am struggling at the moment with running an app ( being developed under Mac OS ) in a unikernel. The app has the [redis-fast-driver](https://github.com/h0x91b/redis-fast-driver) as a dependency.

I see errors regarding incorrect ELF headers.

A short example of the whole workflow of creating a bootable unikernel would be handy.

**Comment by eyberg** [comment]
will need to look into this - i'm not super familiar w/the way node does cross-compilation which is needed for this dep cause it has c bindings but it looks like using node-pre-gyp might be what people use ->

```
node-pre-gyp -C node_modules/redis-fast-driver clean configure --arch=x86_64 --target_platform=linux rebuild
```

the other hacky way of getting around this is to just build that dep on linux and then tx it over, but cross-compilation should def. be possible

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg,

thank you for your message. I decided to move the build step of a node application to docker. Thus, I have the redis-fast-driver compiled for linux. So far so good.

According to the [docs](https://nanovms.gitbook.io/ops/basic_usage#running-a-nodejs-script) I have to use the **load** command to "package and deploy".

**The question is**: is there a way for me to just package my nodejs app along with the node being inside a docker container? I would gladly export the resulting package to the host OS and deploy it from there.

**NB**: I have already done my homework and devised a routine for creating the package with the latest version of NodeJS.

**NB**: Documentation is not detailed enough for such a lame as I am to easily work out my way myself.

**Comment by eyberg** [comment]
this briefly touched on here https://github.com/nanovms/ops/pull/488#issuecomment-612981332

but essentially the disk image is built every-time unless you specify it not to be - for instance using the example from the website ->

```
➜  node_example  vi hi.js
➜  node_example  ops load node_v12.13.0 -p 8083 -a hi.js
[node hi.js]
right before expanding
booting /Users/eyberg/.ops/images/node.img ...
```

the new image is built and stored in ~/.ops/images/node.img - you can then transfer that to wherever as it is a disk image and has everything it needs - you can also rename it via '-i'

```
➜  node_example  ops load node_v12.13.0 -p 8083 -a hi.js -i mynodeapp
[node hi.js]
right before expanding
booting /Users/eyberg/.ops/images/mynodeapp ...
```

if you are wanting to deploy straight to the cloud first you could then do a 

```
ops image create -c config.json -p node_v12.3.0 -a hi.js
```


* any other documentation you feel is missing and you could add on we definitely accept PRs - this repo instantly reflects what is on https://nanovms.gitbook.io/ops/

**Comment by eyberg** [comment]
https://github.com/nanovms/ops/issues/502

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg,

first of all - thank you for your help. a lot!

I am almost there (being inside a docker container with debian:bullseye):
- package for node `14.1.0` is being built and deployed to `.ops/local_packages` without problems
- dependencies of my project (based on [pnpm](https://pnpm.js.org/en/motivation) - w/ hard links to modules in `node_modules` directories) are installed locally without problems
- the project itself tries to start with `ops load -nfl node_v14.1.0 -c ./nanovms.config.json`
- I am planning to write a tutorial using a slightly simplified approach and publish it here as a PR.

but.

there is a problem. as always. Let's first look at the directory layout within a docker container ( a simplified version ) of the project:

```
/install-root
    |
    +-sources
    |     +-libs
    |     +-services
    |         +-authswitch
    |           +-(authswitch.mjs, package.json, .env, nanovms.config.json, node_modules)
    +- ...
```

this is how the unikernel gets started:

```docker
WORKDIR /install-root
RUN ops load -nfl node_v14.1.0 -c /install-root/sources/services/authswitch/nanovms.config.json
```

NB: please note that the `dotenv` module in the `/install-root/sources/services/authswitch/node_modules` directory is a hard link to the `/install-root/node_modules/dotenv`. This is ( my guess ) specific to how `pnpm` works.

Starting the `authswitch` on the host machine (mac os) from the solution root directory proves the `authswitch` works without problems:

```sh
node sources/services/authswitch/authswitch.mjs
```

unfortunately, starting the unikernel from the `/install-root` leads to the following error:

```sh
Step 15/15 : RUN ops load -nfl node_${NODEJS_VERSION} -c /install-root/sources/services/${SERVICE_NAME}/nanovms.config.json
 ---> Running in 6c6c97289dec
Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-linux.tar.gz
 100% |████████████████████████████████████████|  [0s:0s]
[node sources/services/authswitch/authswitch.mjs]
booting /root/.ops/images/node.img ...
assigned: 10.0.2.15
ESM 2: Storing file:///sources/services/authswitch/authswitch.mjs in ModuleMap
ESM 2: Translating StandardModule file:///sources/services/authswitch/authswitch.mjs
ESM 2: Translating BuiltinModule nodejs:path
MODULE 2: load native module nodejs:path
ESM 2: Loading BuiltinModule nodejs:path
ESM 2: Storing nodejs:path in ModuleMap
internal/modules/run_main.js:54
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '/sources/services/authswitch/node_modules/dotenv/' imported from /sources/services/authswitch/authswitch.mjs
    at packageMainResolve (internal/modules/esm/resolve.js:461:9)
    at packageResolve (internal/modules/esm/resolve.js:604:14)
    at moduleResolve (internal/modules/esm/resolve.js:649:14)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:695:13)
    at Loader.resolve (internal/modules/esm/loader.js:97:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:243:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:42:40)
    at link (internal/modules/esm/module_job.js:41:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
exit status 3
```

This looks weird to me. Do you have any thoughts?

**Comment by Dmitry-N-Medvedev** [comment]
UPD: okey, it seems that the node process is started from the `/` directory, when it is expected to be run at least from the `/install-root`

**Comment by eyberg** [comment]
ops doesn't currently grab extra files/dirs from symlinks (although i don't see a reason why we couldn't do something to allow that, would need to think about that)

it seems to not be finding the files in /sources/services/authswitch/node_modules/dotenv - is this directory included in the config.json ?

```
➜  tr  tree
.
├── bad
│   └── bob -> /Users/eyberg/tr/good/bob
├── config.json
├── good
│   └── bob
└── main.js

2 directories, 4 files
➜  tr
```

```
{
  "Dirs": ["good"]
}
```

```
const testFolder = '.';
const fs = require('fs');

fs.readdir(testFolder, (err, files) => {
  files.forEach(file => {
    console.log(file);
  });
});
```

**Comment by Dmitry-N-Medvedev** [comment]
refreshener:
1. the whole process is done in a Docker container (since I need to build a native elf redis driver for NodeJS and I cannot do it from my Mac OS)
2. the project is a monorepo under the control of [pnpm](https://pnpm.js.org/en/motivation)
3. the **/install-root** is the root of the monorepo
4. the **/install-root/sources/services/authswitch** is the directory where a node app lives

the following command (run from within a docker container) would install all the dependencies mentioned anywhere under the **/install-root** directory (including **/install-root/sources/services/authswitch**)

```Dockerfile
WORKDIR /install-root
RUN pnpm --recursive install
```

the following command would print out the directory structure under the **/install-root/sources/services/authswitch**:

```Dockerfile
WORKDIR /
RUN ls -laR /install-root/sources/services/authswitch
```

```bash
/install-root/sources/services/authswitch:
total 352
drwxr-xr-x 1 root root   4096 May  5 14:20 .
drwxr-xr-x 1 root root   4096 May  5 14:20 ..
-rw-r--r-- 1 root root    210 Apr 25 15:36 .env
-rw-r--r-- 1 root root   2263 May  5 14:12 authswitch.mjs
-rw-r--r-- 1 root root    985 May  5 14:20 chdir.js
-rw-r--r-- 1 root root    401 May  5 14:20 nanovms.config.json
drwxr-xr-x 3 root root   4096 May  5 14:20 node_modules
-rw-r--r-- 1 root root 323311 Apr 25 20:17 package-lock.json
-rw-r--r-- 1 root root    492 May  5 13:32 package.json
-rw-r--r-- 1 root root    217 May  5 12:21 tree.js
```

```bash
/install-root/sources/services/authswitch/node_modules:
total 20
drwxr-xr-x 3 root root 4096 May  5 14:20 .
drwxr-xr-x 1 root root 4096 May  5 14:20 ..
drwxr-xr-x 2 root root 4096 May  5 14:20 @stammgast
lrwxrwxrwx 1 root root   82 May  5 14:20 dotenv -> ../../../../node_modules/.pnpm/registry.npmjs.org/dotenv/8.2.0/node_modules/dotenv
lrwxrwxrwx 1 root root  104 May  5 14:20 redis-fast-driver -> ../../../../node_modules/.pnpm/registry.npmjs.org/redis-fast-driver/2.1.5/node_modules/redis-fast-driver
```
```bash
/install-root/sources/services/authswitch/node_modules/@stammgast:
total 8
drwxr-xr-x 2 root root 4096 May  5 14:20 .
drwxr-xr-x 3 root root 4096 May  5 14:20 ..
lrwxrwxrwx 1 root root   36 May  5 14:20 libauthswitch -> ../../../../libs/redis/libauthswitch
```

so far so good.

now let's look into the **nanovms.config.json**:

```json
{
  "Dirs": [
    "/install-root"
  ],
  "Args": [
    "/install-root/sources/services/authswitch/chdir.js"
  ],
  "Env": {
    "Environment": "development",
    "NODE_DEBUG": "*",
    "NODE_DEBUG_NATIVE": "*"
  },
  "NameServer": "8.8.8.8",
  "BaseVolumeSz": "100m",
  "RunConfig": {
    "Ports": [80, 8008],
    "Verbose": true,
    "Memory": "2G",
    "Bridged": true
  },
  "RebootOnExit": false
}
```

the contents of the **/install-root/sources/services/authswitch/chdir.js**:

```js
const fs = require('fs');
const path = require('path');
const {
  inspect,
} = require('util');

const tree = {};

const dtree = async (dir, result) => {
  const isDirectory = fs.statSync(dir).isDirectory();
  const p = path.parse(dir);

  if (isDirectory === true) {
    result[p.base] = [];

    const files = await fs.promises.readdir(dir, {
      withFileTypes: true,
    });
    // eslint-disable-next-line no-restricted-syntax
    for await (const file of files) {
      const fsObjectPath = path.resolve(dir, file.name);

      await dtree(fsObjectPath, result[p.base]);
    }
  }

  if (isDirectory === false) {
    result.push(p.base);
  }
};

dtree(path.resolve('/install-root/sources/services/authswitch/node_modules'), tree).then(() => {
  console.error(
    inspect(tree, {
      showHidden: false,
      depth: null,
      colors: true,
    }),
  );
});
```

now let's run a unikernel:

```Dockerfile

RUN ops load -nfl node_v14.1.0 --config /install-root/sources/services/authswitch/nanovms.config.json -i service:authswitch
```

and here is the result:

```Dockerfile
 ---> Running in ef7aa2d99781
Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-linux.tar.gz
 100% |████████████████████████████████████████|  [0s:0s]
[node /install-root/sources/services/authswitch/chdir.js]
booting /root/.ops/images/service:authswitch ...
assigned: 10.0.2.15
MODULE 2: looking for "/install-root/sources/services/authswitch/chdir.js" in ["/lib/node"]
MODULE 2: load "/install-root/sources/services/authswitch/chdir.js" for module "."
MODULE 2: Module._load REQUEST fs parent: .
MODULE 2: load native module fs
MODULE 2: Module._load REQUEST path parent: .
MODULE 2: load native module path
MODULE 2: Module._load REQUEST util parent: .
MODULE 2: load native module util
ESM 2: Storing file:///install-root/sources/services/authswitch/chdir.js in ModuleMap
{
  node_modules: [
    '@stammgast': [ libauthswitch: [] ],
    dotenv: [],
    'redis-fast-driver': []
  ]
}
exit status 1
Removing intermediate container ef7aa2d99781
 ---> c96ef94e9c9d
Successfully built c96ef94e9c9d
Successfully tagged kaufmann-service-authswitch:latest
```

as we can see:

1. the **node_modules/@stammgast/libauthswitch** is an empty directory in the unikernel (but it is not empty in the docker container)
2. the **node_modules/@stammgast/dotenv** is an empty directory in the unikernel (but it is not empty in the docker container)
3. the **node_modules/@stammgast/redis-fast-driver** is an empty directory in the unikernel (but it is not empty in the docker container)

you can go to the beginning of this message where you can see that the **/install-root/sources/services/authswitch/node_modules** directory within the docker container is not empty.

I might be missing something very-very simple here. Any thoughts?

**Comment by eyberg** [comment]
was trying to replicate what you have and ran into this 

https://github.com/nanovms/nanos/issues/1167

I couldn't immediately repro this cause of that issue, however, the problem you are having is that your symlinks created by pnpm don't look like they are being included into the disk image

from what I was trying to replicate, it does appear that ops is at least telling mkfs to insert the symlinks as desired ->

you can save the manifest via the config.json ->

```
  "ManifestName": "my.manifest",
```

```
children:(
    node_v13.6.0:(children:(
        node:(contents:(host:/home/eyberg/.ops/.staging/node_v13.6.0/node))
        package.manifest:(contents:(host:/home/eyberg/.ops/.staging/node_v13.6.0/package.manifest))
    ))
    kernel:(contents:(host:/home/eyberg/.ops/0.1.25/stage3.img))
    etc:(children:(
        resolv.conf:(contents:(host:/tmp/node_temp185535012/resolv.conf))
        passwd:(contents:(host:/tmp/node_temp185535012/passwd))
    ))
    bob.js:(contents:(host:bob.js))
    pnpm-lock.yaml:(contents:(host:pnpm-lock.yaml))
    node_modules:(children:(
        .modules.yaml:(contents:(host:node_modules/.modules.yaml))
        .pnpm:(children:(
            lock.yaml:(contents:(host:node_modules/.pnpm/lock.yaml))
            node_modules:(children:(
                nan:(children:())
            ))
            registry.npmjs.org:(children:(
                nan:(children:(
                    2.14.1:(children:(
                        node_modules:(children:(
                            nan:(children:(
                                nan.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan.h))
                                nan_implementation_12_inl.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_implementation_12_inl.h))
                                nan_json.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_json.h))
                                nan_persistent_12_inl.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_persistent_12_inl.h))
                                nan_private.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_private.h))
                                README.md:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/README.md))
                                nan_string_bytes.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_string_bytes.h))
                                nan_callbacks.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_callbacks.h))
                                nan_converters_pre_43_inl.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_converters_pre_43_inl.h))
                                include_dirs.js:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/include_dirs.js))
                                nan_callbacks_12_inl.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_callbacks_12_inl.h))
                                nan_maybe_43_inl.h:(contents:(host:node_modules/.pnpm/registry.npmjs.org/nan/2.14.1/node_modules/nan/nan_maybe_43_inl.h))
                                doc:(children:(
```

i'm not a big node user but when I installed pnpm the modules ended up like so:

```
lrwxrwxrwx 1 eyberg eyberg 79 May  5 09:31 redis-fast-driver -> .pnpm/registry.npmjs.org/redis-fast-driver/2.1.5/node_modules/redis-fast-driver
```

your modules, however, are pointing to /install-root or outside of it /.node_modules ? - can you tree from there? '/' ?

**Comment by Dmitry-N-Medvedev** [comment]
quick question: is it possible to run **pnpm --recursive install** command right before the unikernel is built? It would help me ensure all the sym links are in place.

**Comment by eyberg** [comment]
whatever you run would be ran outside so yes, does npm try to install extra packages on-demand?

the thing is in your example

```
/install-root/sources/services/authswitch/node_modules:
total 20
drwxr-xr-x 3 root root 4096 May  5 14:20 .
drwxr-xr-x 1 root root 4096 May  5 14:20 ..
drwxr-xr-x 2 root root 4096 May  5 14:20 @stammgast
lrwxrwxrwx 1 root root   82 May  5 14:20 dotenv -> ../../../../node_modules/.pnpm/registry.npmjs.org/dotenv/8.2.0/node_modules/dotenv
lrwxrwxrwx 1 root root  104 May  5 14:20 redis-fast-driver -> ../../../../node_modules/.pnpm/registry.npmjs.org/redis-fast-driver/2.1.5/node_modules/redis-fast-driver
```

to me that reads that you have (in your container) the following directories:

'/node_modules' && perhaps '/.pnpm' ? from the redis example '/install-root/libs'

we should find out what directory that is actually pointing too - inside your docker container you can just execute 'tree' to confirm but also try throwing in '/node_modules' to your 'Dirs' list in the ops config.json

**Comment by Dmitry-N-Medvedev** [comment]
I have cut off all the directories that reside in **/** with the exception of **/install-root**. Below you will find the contents of the **/install-root** only:

```bash
Step 18/20 : WORKDIR /install-root
 ---> Running in 1de844d56a00
Removing intermediate container 1de844d56a00
 ---> 04b5f59f38f9
Step 19/20 : RUN pnpm --recursive install
 ---> Running in cd553e2a3123
Scope: all 4 workspace projects
Lockfile is up-to-date, resolution step is skipped
Removing intermediate container cd553e2a3123
 ---> 70a4fd89d957
Step 20/20 : RUN tree .
 ---> Running in 213686af5883
.
|-- node_modules
|-- pnpm-lock.yaml
|-- pnpm-workspace.yaml
`-- sources
    |-- libs
    |   |-- fbs
    |   |   |-- bin
    |   |   |   |-- cli.mjs
    |   |   |   `-- lib
    |   |   |       |-- libpathmapper.mjs
    |   |   |       |-- libschemacompiler.mjs
    |   |   |       `-- specs
    |   |   |           |-- libpathmapper.spec.mjs
    |   |   |           `-- libschemacompiler.spec.mjs
    |   |   |-- generated
    |   |   |   `-- web-app
    |   |   |       |-- InternalMessage
    |   |   |       |   |-- Database
    |   |   |       |   |   |-- getTokenMessage.generated.mjs
    |   |   |       |   |   `-- setTokenMessage.generated.mjs
    |   |   |       |   `-- WebWorker
    |   |   |       |       |-- startMessage.generated.mjs
    |   |   |       |       `-- stopMessage.generated.mjs
    |   |   |       `-- InternalMessage.generated.mjs
    |   |   |-- helpers
    |   |   |   `-- InternalMessage
    |   |   |       |-- InternalMessageType.mjs
    |   |   |       |-- getMessageType.mjs
    |   |   |       |-- messageHasType.mjs
    |   |   |       |-- newDatabaseGetTokenMessage.mjs
    |   |   |       |-- newDatabaseSetTokenMessage.mjs
    |   |   |       |-- newWebWorkerStartMessage.mjs
    |   |   |       |-- newWebWorkerStopMessage.mjs
    |   |   |       `-- specs
    |   |   |           `-- InternalMessage.spec.mjs
    |   |   |-- node_modules
    |   |   |   |-- babel-eslint -> ../../../../node_modules/.pnpm/registry.npmjs.org/babel-eslint/10.1.0_eslint@6.8.0/node_modules/babel-eslint
    |   |   |   |-- chai -> ../../../../node_modules/.pnpm/registry.npmjs.org/chai/4.2.0/node_modules/chai
    |   |   |   |-- commander -> ../../../../node_modules/.pnpm/registry.npmjs.org/commander/5.1.0/node_modules/commander
    |   |   |   |-- coveralls -> ../../../../node_modules/.pnpm/registry.npmjs.org/coveralls/3.1.0/node_modules/coveralls
    |   |   |   |-- cross-env -> ../../../../node_modules/.pnpm/registry.npmjs.org/cross-env/7.0.2/node_modules/cross-env
    |   |   |   |-- eslint -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint/6.8.0/node_modules/eslint
    |   |   |   |-- eslint-config-airbnb-base -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-config-airbnb-base/14.1.0_8cdb6d8c18c3319a1365bd5afa0063a3/node_modules/eslint-config-airbnb-base
    |   |   |   |-- eslint-plugin-chai-friendly -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-plugin-chai-friendly/0.5.0_eslint@6.8.0/node_modules/eslint-plugin-chai-friendly
    |   |   |   |-- eslint-plugin-import -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-plugin-import/2.20.2_eslint@6.8.0/node_modules/eslint-plugin-import
    |   |   |   |-- eslint-plugin-node -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-plugin-node/11.1.0_eslint@6.8.0/node_modules/eslint-plugin-node
    |   |   |   |-- execa -> ../../../../node_modules/.pnpm/registry.npmjs.org/execa/4.0.0/node_modules/execa
    |   |   |   |-- flatbuffers -> ../../../../node_modules/.pnpm/registry.npmjs.org/flatbuffers/1.12.0/node_modules/flatbuffers
    |   |   |   |-- form-data -> ../../../../node_modules/.pnpm/registry.npmjs.org/form-data/3.0.0/node_modules/form-data
    |   |   |   |-- got -> ../../../../node_modules/.pnpm/registry.npmjs.org/got/10.7.0/node_modules/got
    |   |   |   |-- mocha -> ../../../../node_modules/.pnpm/registry.npmjs.org/mocha/7.1.2/node_modules/mocha
    |   |   |   |-- nanoid -> ../../../../node_modules/.pnpm/registry.npmjs.org/nanoid/3.1.4/node_modules/nanoid
    |   |   |   |-- nyc -> ../../../../node_modules/.pnpm/registry.npmjs.org/nyc/15.0.1/node_modules/nyc
    |   |   |   `-- which -> ../../../../node_modules/.pnpm/registry.npmjs.org/which/2.0.2/node_modules/which
    |   |   |-- package.json
    |   |   `-- schema
    |   |       `-- web-app
    |   |           |-- InternalMessage
    |   |           |   |-- Database
    |   |           |   |   |-- getTokenMessage.fbs
    |   |           |   |   `-- setTokenMessage.fbs
    |   |           |   `-- WebWorker
    |   |           |       |-- startMessage.fbs
    |   |           |       `-- stopMessage.fbs
    |   |           `-- InternalMessage.fbs
    |   |-- pack-project
    |   |   |-- bin
    |   |   |   |-- cli.mjs
    |   |   |   `-- lib
    |   |   |       |-- libpathmapper.mjs
    |   |   |       |-- libschemacompiler.mjs
    |   |   |       `-- specs
    |   |   |           |-- libpathmapper.spec.mjs
    |   |   |           `-- libschemacompiler.spec.mjs
    |   |   |-- node_modules
    |   |   |   |-- chai -> ../../../../node_modules/.pnpm/registry.npmjs.org/chai/4.2.0/node_modules/chai
    |   |   |   |-- commander -> ../../../../node_modules/.pnpm/registry.npmjs.org/commander/5.1.0/node_modules/commander
    |   |   |   |-- cross-env -> ../../../../node_modules/.pnpm/registry.npmjs.org/cross-env/7.0.2/node_modules/cross-env
    |   |   |   |-- eslint -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint/6.8.0/node_modules/eslint
    |   |   |   |-- eslint-config-airbnb-base -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-config-airbnb-base/14.1.0_8cdb6d8c18c3319a1365bd5afa0063a3/node_modules/eslint-config-airbnb-base
    |   |   |   |-- eslint-plugin-chai-friendly -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-plugin-chai-friendly/0.6.0_eslint@6.8.0/node_modules/eslint-plugin-chai-friendly
    |   |   |   |-- eslint-plugin-import -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-plugin-import/2.20.2_eslint@6.8.0/node_modules/eslint-plugin-import
    |   |   |   |-- eslint-plugin-node -> ../../../../node_modules/.pnpm/registry.npmjs.org/eslint-plugin-node/11.1.0_eslint@6.8.0/node_modules/eslint-plugin-node
    |   |   |   |-- mocha -> ../../../../node_modules/.pnpm/registry.npmjs.org/mocha/7.1.2/node_modules/mocha
    |   |   |   `-- nyc -> ../../../../node_modules/.pnpm/registry.npmjs.org/nyc/15.0.1/node_modules/nyc
    |   |   `-- package.json
    |   `-- redis
    |       `-- libauthswitch
    |           |-- libauthswitch.mjs
    |           |-- node_modules
    |           |   |-- chai -> ../../../../../node_modules/.pnpm/registry.npmjs.org/chai/4.2.0/node_modules/chai
    |           |   |-- coveralls -> ../../../../../node_modules/.pnpm/registry.npmjs.org/coveralls/3.1.0/node_modules/coveralls
    |           |   |-- cross-env -> ../../../../../node_modules/.pnpm/registry.npmjs.org/cross-env/7.0.2/node_modules/cross-env
    |           |   |-- eslint -> ../../../../../node_modules/.pnpm/registry.npmjs.org/eslint/6.8.0/node_modules/eslint
    |           |   |-- mocha -> ../../../../../node_modules/.pnpm/registry.npmjs.org/mocha/7.1.2/node_modules/mocha
    |           |   |-- nanoid -> ../../../../../node_modules/.pnpm/registry.npmjs.org/nanoid/3.1.4/node_modules/nanoid
    |           |   |-- nyc -> ../../../../../node_modules/.pnpm/registry.npmjs.org/nyc/15.0.1/node_modules/nyc
    |           |   `-- redis-fast-driver -> ../../../../../node_modules/.pnpm/registry.npmjs.org/redis-fast-driver/2.1.5/node_modules/redis-fast-driver
    |           |-- package.json
    |           `-- specs
    |               `-- libauthswitch.spec.mjs
    `-- services
        `-- authswitch
            |-- authswitch.mjs
            |-- chdir.js
            |-- nanovms.config.json
            |-- node_modules
            |   |-- @stammgast
            |   |   `-- libauthswitch -> ../../../../libs/redis/libauthswitch
            |   |-- dotenv -> ../../../../node_modules/.pnpm/registry.npmjs.org/dotenv/8.2.0/node_modules/dotenv
            |   `-- redis-fast-driver -> ../../../../node_modules/.pnpm/registry.npmjs.org/redis-fast-driver/2.1.5/node_modules/redis-fast-driver
            |-- package-lock.json
            |-- package.json
            `-- tree.js

73 directories, 41 files
```

**Comment by Dmitry-N-Medvedev** [comment]
PS: the **/install-root/node_modules** is empty. Weird. Need some time to investigate it.

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg,

I have updated my previous message after building a tree w/ all the hidden files:

```Dockerfile
RUN tree -an .
```

so, as far as I can tell from the output, all the necessary pnpm related directories are located within the **/install-root** directory.

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg,

quick question: you mentioned that "you can save the manifest via the config.json". Since I'm a nub here I am curious if this is a requirement?

**Comment by eyberg** [comment]
no it's not mandatory - strictly for debugging what gets put on the filesystem,

also I saw you were adding in  ```"Bridged": true``` <-- that's if you wanted to create a tap and add it to a network bridge - most ppl prob shouldn't use this unless they are on bare-metal linux

the "BaseVolumeSz" - that's also an optional flag to increase the disk size if you need additional space (eg: for a database) - most applications don't need it out of the box and in this case I'd ensure that building without that option is not a larger image than with, the build log looks relatively large

so basically comparing the tree output inside docker to the tree inside of the unikernel should tell you what is missing or in the wrong place, once https://github.com/nanovms/nanos/issues/1167 is fixed this should be easier for me to repro

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg,

thank you :)

can't wait for the issue mentioned to be fixed.

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg,

I guess the title **please add NodeJS project w/ a dependency that has native module** is kind of inaccurate now.

A better name would be something like **NodeJS with dependencies as hard links (pnpm)**.

What do you think?

**Comment by eyberg** [comment]
https://github.com/nanovms/nanos/issues/1169

**Comment by eyberg** [comment]
ok - so I found the original problem was what we were speculating w/pnpm - ops actually ignores symlinks in various places right now 

```
+++ b/lepton/manifest.go
@@ -91,7 +91,7 @@ func (m *Manifest) AddDirectory(dir string) error {
                        vmpath = hostpath
                }

-               if (info.Mode() & os.ModeSymlink) != 0 {
+/*             if (info.Mode() & os.ModeSymlink) != 0 {
```

however, that's because we have in the past supported only guest symlinks not host symlinks - it's kinda up in the air how we want to go about resolving those through mkfs - but https://github.com/nanovms/nanos/issues/1169 should mediate that

if you can build with regular npm you might be able to get around this for now (i was able too), otherwise will need to hang tight on 1169

it should be noted though that both the node version && whatever system libc you have should match

**Comment by Dmitry-N-Medvedev** [comment]
> if you can build with regular npm you might be able to get around this for now

I will proceed programming my system w/ pnpm for now, waiting for this issue to eventually resolve

> it should be noted though that both the node version && whatever system libc you have should match

I am building node package using **ldd node**'s output. My guess it is enough to ensure libc is listed also, isn't it?

**Comment by eyberg** [comment]
if you are building your own node pkg you should be ok on the addon yes

**Comment by eyberg** [comment]
w/latest ops 0.1.10 and using the new node_v14.2.0 pkg I was able to get pnpm working w/the redis addon via https://github.com/nanovms/nanos/issues/1169 , lmk if you run into anything else

**Comment by Dmitry-N-Medvedev** [comment]
hi @eyberg,

this is what I bumped into today:

```bash
Step 18/18 : RUN ops load -nfl node_${NODEJS_VERSION} --config /install-root/sources/services/${SERVICE_NAME}/nanovms.config.json -i service:${SERVICE_NAME}
 ---> Running in 38c5cb652266
Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-linux.tar.gz
 100% |████████████████████████████████████████|  [0s:0s]
[node --trace-warnings /install-root/sources/services/authswitch/authswitch.mjs]
booting /root/.ops/images/service:authswitch ...
qemu-system-x86_64: cannot bind memory to host NUMA nodes: Operation not permitted
exit status 1
```

**Comment by eyberg** [comment]
building them inside of a container is one thing but these really are not made to run inside of a docker container and I don't see it as something that we'd want to support

**Comment by Dmitry-N-Medvedev** [comment]
@eyberg , fair enough :)

I build it via:

```bash
RUN ops image create --config /install-root/sources/services/${SERVICE_NAME}/nanovms.config.json --imagename service:${SERVICE_NAME} --nightly --package node_${NODEJS_VERSION} --target-cloud onprem
```

it runs OK, and produces the following file:

```bash
Step 19/20 : RUN ls -laR /root/.ops/images
 ---> Running in dc8bd8c6d970
/root/.ops/images:
total 113220
drwxr-xr-x 1 root root      4096 May 13 19:02 .
drwxr-xr-x 1 root root      4096 May 13 19:02 ..
-rw-r--r-- 1 root root 115927552 May 13 19:02 service:authswitch
```

a quick question: shouldn't the service:authswitch file have an extension? which kind of image is this?

PS: I probably missed it written somewhere in the docs of course.

**Comment by eyberg** [comment]
yeh - the naming conventions need to be worked out - it is indeed your finished image file - you should be able to copy that out of your container && run it locally w/ops on a mac w/out anything else https://github.com/nanovms/ops/issues/517
